src/Renderer/UI/SimulationView.fs

```diff
@@ -378,23 +384,148 @@ let private viewStatefulComponents step comps numBase model dispatch =
+    let sheetDispatch sMsg = dispatch <| Sheet sMsg
+    let busWireDispatch bMsg = sheetDispatch <| SheetT.Msg.Wire bMsg
+    let symbolDispatch symMsg = busWireDispatch <| BusWireT.Msg.Symbol symMsg
+
+    let tryGetComponentById compId =
+        comps
+        |> List.tryFind (fun comp -> comp.Id = compId)
+    
+    let tryGetConnectionById connId =
+        conns
+        |> List.tryFind (fun conn -> conn.Id = connId)
+    
+    let reacListOfCompsAffected =
+        simError.ComponentsAffected
+        |> List.collect (fun (ComponentId s) -> Option.toList (tryGetComponentById s))
+        |> List.map (fun comp -> li [] [str comp.Label])
+
+    let error =
+        match simError.ErrType with
+        | OutputNotConnected ->
+            let addNCAndRemovePorts() =
+                let addNCtoComponent comp =
+                    comp.OutputPorts
+                    |> List.filter (fun port ->
+                        (conns
+                        |> List.forall (fun conn -> conn.Source.Id <> port.Id))
+                            && (match comp.Type with
+                                | NbitsAdder w ->
+                                    model.Sheet.ChangeAdderComp sheetDispatch (ComponentId comp.Id) (NbitsAdderNoCout w)
+                                    port.PortNumber = Some 0
+                                | NbitsAdderNoCin w ->
+                                    model.Sheet.ChangeAdderComp sheetDispatch (ComponentId comp.Id) (NbitsAdderNoCinCout w)
+                                    port.PortNumber = Some 0
+                                | _ -> true))
+                    |> List.iter (fun port ->
+                        busWireDispatch <| BusWireT.AddNotConnected ((ModelHelpers.tryGetLoadedComponents model), port, {X = comp.X + comp.W * 1.3; Y = comp.Y + comp.H/2.}))
+                        
+                simError.ComponentsAffected
+                |> List.map (fun (ComponentId compId) -> tryGetComponentById compId)
+                |> List.collect Option.toList
+                |> List.iter addNCtoComponent
+                endSimulation endDispatch dispatch
             div [] [
                 str simError.Msg
                 br []
-                str <| "Please fix the error and retry."
+                ul [] reacListOfCompsAffected
+                Button.button [
+                    Button.Color IsSuccess
+                    Button.OnClick (fun _ -> addNCAndRemovePorts())
+                ] [ str "Add 'Not Connected' components and/or remove ports" ]
             ]
-        | Some dep ->
+        | InputNotConnected ->
+            let removeInPorts() =
+                simError.ComponentsAffected
+                |> List.collect (fun (ComponentId compId) -> Option.toList (tryGetComponentById compId))
+                |> List.iter (fun comp ->
+                    match comp.Type with
```

Comment:
factor this to:
function that returns the relevant ports
function that filters then (e.g. not connected or something)
function that changes stuff.
For chnaging component (to remove port) give yourself a function which does the change Component -> Component or even ComponentType -> ComponentType. Not sure if removing these ports ever changes other port numbers (hope not - but need to check - it might). In which case those cases need more work.

Overall - you can use functional abstraction to simplify this stuff.